### PR TITLE
github-ci: add sv live tests - v3


### PR DESCRIPTION
**Pull request by jasonish**
SV_BRANCH=https://github.com/OISF/suricata-verify/pull/3378

Should cover most of the live tests we do in `.github`. DPDK tests that use the
*null* driver as they aren't really applicable to the live scenarios.

TL;DR: I think this covers all the live tests we do in this repo, except the DPDK NULL driver test which doesn't really make sense in the live testing framework.

The following is an AI generated document attempt to show the parity between
the new SV live tests and what we have in `.github`.

# Live-test coverage compared with Suricata main

The live suite covers **36 of the 38 scenarios** run by Suricata main's
`qa/live` scripts. The two excluded scenarios use the DPDK null driver to
generate traffic. The new DPDK tests use controlled ICMP traffic through
AF_PACKET virtual devices, including DPDK bonds.

There are **50 live tests** in total: 36 corresponding to main's scenarios and
14 additional tests. The table below summarizes the coverage; the sections
that follow explain the behavioral differences and map each test to its source.

| Test family in main | Main scenarios | Live tests |
| --- | ---: | ---: |
| AF_PACKET IDS | 4 | 4 |
| Libpcap IDS | 2 | 2 |
| Tenant management | 1 | 1 |
| AF_PACKET/NFQUEUE IPS | 4 | 4 |
| AF_PACKET with Linux bonds | 2 | 2 |
| AF_PACKET/NFQUEUE firewall | 2 | 2 |
| DPDK single-interface configuration | 13 | 13 |
| DPDK bond configuration | 8 | 8 |
| DPDK null-driver smoke tests | 2 | Excluded |
| **Total** | **38** | **36** |

## Differences to keep in mind

The scenarios are covered, with these differences in how they run:

- **Shutdown:** All 15 main IDS, IPS, tenant and firewall scenarios request
  shutdown through the Unix socket. Three corresponding live tests do this;
  the other 12 use SIGTERM. The additional DPDK IDS test also uses socket
  shutdown. NFQUEUE socket shutdown is not yet covered by the live suite.
- **HTTP versus HTTPS:** The four IPS and two Linux-bond tests verify local
  HTTP transfers. Main's scripts attempt HTTPS transfers.
- **Uptime:** The six AF_PACKET/libpcap IDS tests check that uptime is
  nonnegative. Main requires at least 15 seconds.
- **DPDK drivers:** Configuration checks retain main's expected results, but
  use AF_PACKET virtual PMDs instead of null PMDs. These tests exercise virtual
  devices; they do not test physical NICs or VFIO.
- **CI builds:** Main runs live tests under ASAN, coverage instrumentation and
  several DPDK versions. The live workflow here does not yet reproduce that
  matrix. See [CI coverage](#ci-coverage).

## IDS, IPS, tenant and firewall tests

Source paths are relative to main's `qa/live/` directory. The tables omit YAML
arguments and show how each live test shuts Suricata down. Traffic and addressing
are adapted to isolated network namespaces; IDS tests use a local endpoint in
place of the host's default gateway.

### IDS and tenant management

The IDS tests check capture, alerts, dataset clearing, rule reload, socket
queries and hostbits. Libpcap tests also check IPv4-mapped IPv6 dataset lookups
and malformed input rejection. The tenant test checks registration, reload,
unregistration and rejection of an unknown tenant.

| Main script and arguments | Live test | Shutdown |
| --- | --- | --- |
| [afp-ids.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/afp-ids.sh) `2 autofp` | [afp-ids-tpacket2-autofp](tests/afp-ids-tpacket2-autofp/test.yaml) | SIGTERM |
| [afp-ids.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/afp-ids.sh) `2 workers` | [afp-ids-tpacket2-workers](tests/afp-ids-tpacket2-workers/test.yaml) | Socket |
| [afp-ids.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/afp-ids.sh) `3 autofp` | [afp-ids-tpacket3-autofp](tests/afp-ids-tpacket3-autofp/test.yaml) | SIGTERM |
| [afp-ids.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/afp-ids.sh) `3 workers` | [afp-ids-tpacket3-workers](tests/afp-ids-tpacket3-workers/test.yaml) | SIGTERM |
| [pcap.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/pcap.sh) `autofp` | [pcap-ids-autofp](tests/pcap-ids-autofp/test.yaml) | SIGTERM |
| [pcap.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/pcap.sh) `single` | [pcap-ids-single](tests/pcap-ids-single/test.yaml) | Socket |
| [multi-tenant.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/multi-tenant.sh) `autofp` | [pcap-multi-tenant-autofp](tests/pcap-multi-tenant-autofp/test.yaml) | Socket |

The AF_PACKET names follow main's script arguments, which are inverted:
argument `2` enables `tpacket-v3`, while `3` disables it. Both capture settings
and both runmodes are covered.

### IPS, Linux bonds and firewall

The IPS tests check that web traffic passes and ICMP is blocked, using counters
and a server-side packet capture. The Linux-bond tests also cover fragmented
ICMP at MTU 9000; the two-network case checks drops and per-source counts across
both networks. Firewall tests cover default drop, curl blocking, wget
acceptance, rule reload, alerts, counters and server-side capture.

| Main script and arguments | Live test | Shutdown |
| --- | --- | --- |
| [netns/afp-ips-netns-bridge.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/afp-ips-netns-bridge.sh) `2 workers` | [ips-drop-icmp-afp](tests/ips-drop-icmp-afp/test.yaml) | SIGTERM |
| [netns/afp-ips-netns-bridge.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/afp-ips-netns-bridge.sh) `2 autofp` | [ips-drop-icmp-afp-autofp](tests/ips-drop-icmp-afp-autofp/test.yaml) | SIGTERM |
| [netns/nfq-ips-netns-route.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/nfq-ips-netns-route.sh) `autofp` | [ips-drop-icmp-nfq](tests/ips-drop-icmp-nfq/test.yaml) | SIGTERM |
| [netns/nfq-ips-netns-route.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/nfq-ips-netns-route.sh) `workers` | [ips-drop-icmp-nfq-workers](tests/ips-drop-icmp-nfq-workers/test.yaml) | SIGTERM |
| [netns/afp-ips-netns-bond-bridge.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/afp-ips-netns-bond-bridge.sh) `2 workers` | [afp-ips-bond](tests/afp-ips-bond/test.yaml) | SIGTERM |
| [netns/afp-ips-netns-bond-bridge2.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/afp-ips-netns-bond-bridge2.sh) `2 workers` | [afp-ips-bond-two-networks](tests/afp-ips-bond-two-networks/test.yaml) | SIGTERM |
| [netns/afp-fw-netns-bridge.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/afp-fw-netns-bridge.sh) `2 workers` | [afp-fw](tests/afp-fw/test.yaml) | SIGTERM |
| [netns/nfq-fw-netns-route.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/netns/nfq-fw-netns-route.sh) `autofp` | [nfq-fw](tests/nfq-fw/test.yaml) | SIGTERM |

## DPDK configuration tests

All **21 cases** from main's [DPDK configuration suite](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/dpdk/dpdk-testsuite.sh)
are included: 14 successful starts and seven expected startup failures. They
retain the original configuration overrides and expected pool/cache sizes or
error messages.

Successful tests also require an ICMP alert and a clean shutdown with exit
status 0. Expected startup failures must exit with status 1 before the engine
becomes ready. Main's [helper](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/dpdk/dpdk-checklog.sh)
accepts any nonzero startup failure and ignores the shutdown exit status.

The bond tests exercise DPDK's `net_bonding0` with two AF_PACKET PMD members.
These are separate from the Linux bonds used by the IPS tests above. Successful
configuration tests use SIGTERM for shutdown. Tests skip when the required CPU
capacity or DPDK bond support is unavailable.

### Single-interface cases

Pool sizes below are per pool.

| Main configuration | Live test | Expected result |
| --- | --- | --- |
| IDS: auto mempool, auto cache (1 thread) | [dpdk-config-ids-auto-auto-1](tests/dpdk-config-ids-auto-auto-1/test.yaml) | Starts: 1 pool of 63, cache 21. |
| IDS: auto mempool, auto cache (2 threads) | [dpdk-config-ids-auto-auto-2](tests/dpdk-config-ids-auto-auto-2/test.yaml) | Starts: 2 pools of 63, cache 21. |
| IDS: auto mempool, static cache=1 (1 thread) | [dpdk-config-ids-auto-cache1-1](tests/dpdk-config-ids-auto-cache1-1/test.yaml) | Starts: 1 pool of 63, cache 1. |
| IDS: auto mempool, static cache=1 (2 threads) | [dpdk-config-ids-auto-cache1-2](tests/dpdk-config-ids-auto-cache1-2/test.yaml) | Starts: 2 pools of 63, cache 1. |
| IDS: auto mempool, oversized cache=1024 (fail) | [dpdk-config-ids-cache-too-large](tests/dpdk-config-ids-cache-too-large/test.yaml) | Fails: invalid cache size. |
| IDS: static mempool=1023, auto cache (1 thread) | [dpdk-config-ids-pool1023-1](tests/dpdk-config-ids-pool1023-1/test.yaml) | Starts: 1 pool of 1023, cache 341. |
| IDS: static mempool=1024, auto cache (1 thread) | [dpdk-config-ids-pool1024-rounding](tests/dpdk-config-ids-pool1024-rounding/test.yaml) | Starts: 1 pool of 1023, cache 341. |
| IDS: static mempool=1023, auto cache (2 threads) | [dpdk-config-ids-pool1023-2](tests/dpdk-config-ids-pool1023-2/test.yaml) | Starts: 2 pools of 511, cache 73. |
| IDS: mempool too small (fail) | [dpdk-config-ids-pool-too-small](tests/dpdk-config-ids-pool-too-small/test.yaml) | Fails: pool too small for descriptors/queues. |
| IDS: mempool too small for queues (fail) | [dpdk-config-ids-pool-too-small-queues](tests/dpdk-config-ids-pool-too-small-queues/test.yaml) | Fails: pool too small for descriptors/queues. |
| IDS: static mempool=32, power-of-two boundary (fail) | [dpdk-config-ids-pool32-boundary](tests/dpdk-config-ids-pool32-boundary/test.yaml) | Fails: pool too small for descriptors/queues. |
| IDS: static mempool=64, power-of-two boundary (1 thread) | [dpdk-config-ids-pool64-boundary](tests/dpdk-config-ids-pool64-boundary/test.yaml) | Starts: 1 pool of 63, cache 21. |
| IDS: auto descriptors, auto mempool (OOM expected with nohuge) | [dpdk-config-ids-auto-descriptors](tests/dpdk-config-ids-auto-descriptors/test.yaml) | Fails: pool 65535/cache 257 logged, then allocation failure. |

### Bond cases

| Main configuration | Live test | Expected result |
| --- | --- | --- |
| Bond: auto mempool, auto cache (1 thread) | [dpdk-config-bond-auto-auto-1](tests/dpdk-config-bond-auto-auto-1/test.yaml) | Starts: 1 pool of 127, cache 1. |
| Bond: auto mempool, auto cache (2 threads) | [dpdk-config-bond-auto-auto-2](tests/dpdk-config-bond-auto-auto-2/test.yaml) | Starts: 2 pools of 127, cache 1. |
| Bond: auto mempool, static cache=7 (1 thread) | [dpdk-config-bond-auto-cache7-1](tests/dpdk-config-bond-auto-cache7-1/test.yaml) | Starts: 1 pool of 127, cache 7. |
| Bond: auto mempool, static cache=7 (2 threads) | [dpdk-config-bond-auto-cache7-2](tests/dpdk-config-bond-auto-cache7-2/test.yaml) | Starts: 2 pools of 127, cache 7. |
| Bond: auto mempool, oversized cache=1024 (fail) | [dpdk-config-bond-cache-too-large](tests/dpdk-config-bond-cache-too-large/test.yaml) | Fails: invalid cache size. |
| Bond: static mempool=1023, auto cache (1 thread) | [dpdk-config-bond-pool1023-1](tests/dpdk-config-bond-pool1023-1/test.yaml) | Starts: 1 pool of 1023, cache 341. |
| Bond: static mempool=1023, auto cache (2 threads) | [dpdk-config-bond-pool1023-2](tests/dpdk-config-bond-pool1023-2/test.yaml) | Starts: 2 pools of 511, cache 73. |
| Bond: mempool too small (fail) | [dpdk-config-bond-pool-too-small](tests/dpdk-config-bond-pool-too-small/test.yaml) | Fails: pool too small for descriptors/queues. |

### Excluded null-driver tests

Main's [dpdk.sh](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/qa/live/dpdk.sh) runs IDS and IPS smoke tests using traffic
generated by the null PMD. Both are intentionally excluded.

[dpdk-ids-afpacket-pmd](tests/dpdk-ids-afpacket-pmd/test.yaml) provides DPDK IDS
capture, rule reload and socket-shutdown coverage with controlled traffic and
alert checks. A DPDK inline IPS traffic test remains a possible addition.

## Additional live tests

These 14 tests add coverage beyond main's `qa/live` scenarios. The comparison
is limited to those scripts; it does not include main's unit tests or the
regular Suricata-Verify packet-replay suite.

| Area | Additional tests |
| --- | --- |
| Basic HTTP through IDS and IPS | [simple-http-ids](tests/simple-http-ids/test.yaml), [simple-http-afp](tests/simple-http-afp/test.yaml), [simple-http-nfq](tests/simple-http-nfq/test.yaml) |
| Static IPv6 IDS traffic | [simple-icmpv6-ids](tests/simple-icmpv6-ids/test.yaml) |
| DPDK IDS with controlled traffic, reload and socket shutdown | [dpdk-ids-afpacket-pmd](tests/dpdk-ids-afpacket-pmd/test.yaml) |
| Capture/offload bypass with end-to-end traffic | [afp-ips-xdp-bypass](tests/afp-ips-xdp-bypass/test.yaml), [nfq-ips-bypass](tests/nfq-ips-bypass/test.yaml) |
| Local IPS bypass | [ips-local-bypass-afp](tests/ips-local-bypass-afp/test.yaml), [ips-local-bypass-nfq](tests/ips-local-bypass-nfq/test.yaml) |
| Overlapping alert/drop rules | [ips-alert-drop-icmp-afp](tests/ips-alert-drop-icmp-afp/test.yaml), [ips-alert-drop-http-afp](tests/ips-alert-drop-http-afp/test.yaml) |
| Firewall default drop, then ICMP acceptance after reload | [firewall-icmp](tests/firewall-icmp/test.yaml) |
| Firewall response-body rejection | [ruletype-firewall-96-lt-response-body-no-match-afp](tests/ruletype-firewall-96-lt-response-body-no-match-afp/test.yaml), [ruletype-firewall-96-lt-response-body-no-match-nfq](tests/ruletype-firewall-96-lt-response-body-no-match-nfq/test.yaml) |

## CI coverage

Main currently invokes its existing scripts in these four jobs. Repeating a
scenario across jobs does not increase the scenario count above.

| Main job | Tests run | Build |
| --- | --- | --- |
| [Ubuntu 24.04: capture coverage](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/.github/workflows/builds.yml#L1755) | AF_PACKET IDS, libpcap IDS, tenants, DPDK configuration/bonds and null-PMD smoke tests. | Clang 19 C/C++ coverage and Rust coverage; no ASAN. |
| [Ubuntu 24.04: capture ASAN](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/.github/workflows/builds.yml#L2153) | Same tests as the capture coverage job, including DPDK bonds. | Clang 18 with AddressSanitizer; packaged DPDK. |
| [Ubuntu 24.04: namespace coverage](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/.github/workflows/builds.yml#L1996) | AF_PACKET/NFQUEUE IPS, Linux bonds and firewall. | Clang 19 C/C++ coverage and Rust coverage; no ASAN. |
| [Ubuntu 22.04: DPDK versions](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/.github/workflows/builds.yml#L2841) | DPDK configuration/bonds and null-PMD smoke tests. | DPDK 21.11.9, 22.11.10, 23.11.5 and 24.11.3; no ASAN or coverage instrumentation. |

The ASAN build uses
`CFLAGS="-g -fsanitize=address -fno-omit-frame-pointer"`.

The [live workflow](../.github/workflows/live.yaml) here runs in a privileged
Ubuntu 26.04 container. It builds Suricata with debug, NFQUEUE, eBPF and DPDK
support, then runs `live/run.py`. It has no ASAN, coverage instrumentation or
DPDK-version matrix yet.

There are separate ASAN jobs for the [regular verification suite](../.github/workflows/builds.yml)
and main's [pcap Unix-socket tests](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/.github/workflows/builds.yml#L1906).
Those run `run.py` and `qa/unix.sh`, respectively, and do not provide ASAN
coverage for this live suite.

## Validation and comparison basis

Local validation completed so far:

- All 21 DPDK configuration tests passed without skips using DPDK 25.11.1.
- All four tests using client-requested socket shutdown passed, including
  shutdown acknowledgement and clean exit.
- Those runs used an ASAN build with leak detection disabled.

Older DPDK versions and remote CI have not been validated for these ports.
The coverage comparison is based on source and assertion checks; it does not
establish a successful run of all 50 tests on every platform.

Compared on 2026-09-21 against Suricata main
[`9b7938aec3`](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d) and live-suite revision `9ff23d63`.
The scenario inventory comes from main's [build workflow](https://github.com/OISF/suricata/blob/9b7938aec3c66cf4b467d281069b913b586bdf8d/.github/workflows/builds.yml),
counting each invocation once and expanding the DPDK configuration suite into
its 21 cases.

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA skipped due to no C or rust code changed detected. Set  to force a run. 

Pipeline = code

File: .github/workflows/builds.yml [hunk 1]
```diff
@@ -2944,6 +3021,19 @@ jobs:
           ./qa/live/dpdk.sh "qa/live/dpdk/suricata-null-ips.yaml"
       # DPDK configuration checks
       - run: ./qa/live/dpdk/dpdk-testsuite.sh
+      - name: Run suricata-verify live tests
+        id: sv_live
+        run: ASAN_OPTIONS=detect_leaks=0 python3 ./suricata-verify/live/run.py
```
[the remark this answers is not in the document]
Comment (jasonish) [reply .github/workflows/builds.yml]: This makes sense.

File: .github/workflows/builds.yml [hunk 2]
```diff
@@ -2135,7 +2162,24 @@ jobs:
           LLVM_PROFILE_FILE: "/tmp/afp-fw-netns-bridge.profraw"
       - run: llvm-profdata-19 merge -o afp-fw-netns-bridge.profdata /tmp/afp-fw-netns-bridge.profraw
 
-      - run: llvm-profdata-19 merge -o combined.profdata afp-ips.profdata nfq-ips.profdata afp-ips-autofp.profdata nfq-ips-workers.profdata afp-ips-bond1.profdata afp-ips-bond2.profdata afp-fw-netns-bridge.profdata nfq-fw-netns-route.profdata
+      - name: Run suricata-verify live tests
+        id: sv_live
+        run: ASAN_OPTIONS=detect_leaks=0 python3 ./suricata-verify/live/run.py
```
[the remark this answers is not in the document]
Comment (jasonish) [reply .github/workflows/builds.yml:2167]: Makes sense.

**Comment by jasonish** [comment]
I think this test might be a little flaky:
```
===> [pcap-multi-tenant-autofp] Setting up tap environment
===> [pcap-multi-tenant-autofp] Running before script
===> [pcap-multi-tenant-autofp] Starting Suricata
===> [pcap-multi-tenant-autofp] Running Client
===> [pcap-multi-tenant-autofp] Stopping Suricata
===> [pcap-multi-tenant-autofp] FAIL ❌
  client script exited with code 1
  client stdout: /__w/suricata/suricata/suricata-verify/live/tests/pcap-multi-tenant-autofp/output/client.stdout
  client stderr: /__w/suricata/suricata/suricata-verify/live/tests/pcap-multi-tenant-autofp/output/client.stderr
  ```

**Comment by suricata-qa** [comment]
Information: QA skipped due to no C or rust code changed detected. Set  to force a run. 

Pipeline = code

**Comment by suricata-qa** [comment]
Information: QA skipped due to no C or rust code changed detected. Set  to force a run. 

Pipeline = code

**Comment by suricata-qa** [comment]
Information: QA skipped due to no C or rust code changed detected. Set  to force a run. 

Pipeline = code

**Comment by jasonish** [comment]
@suricata-review-bot Are all the review comments addressed?

**Comment by lukashino** [comment]
Just to comment on DPDK NULL test

> TL;DR: I think this covers all the live tests we do in this repo, except the DPDK NULL driver test which doesn't really make sense in the live testing framework.

The NULL tests are smoke tests to verify that we can actually run Suricata with DPDK.
The work you did already covers that through net_af_packet tests, so we can ignore and drop the tests.

**Comment by jasonish** [comment]
Replaced by https://github.com/OISF/suricata/pull/16284
